ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala

```diff
@@ -243,6 +262,86 @@ object KeyValueConsumption {
     )
   }
 
+  private[kvutils] def outOfTimeBoundsEntryToUpdate(
+      recordTime: Timestamp,
+      outOfTimeBoundsEntry: DamlOutOfTimeBoundsEntry): Option[Update] = {
+    val wrappedLogEntry = outOfTimeBoundsEntry.getEntry
+    val duplicateUntilMaybe = parseOptionalTimestamp(
+      outOfTimeBoundsEntry.hasDuplicateUntil,
+      outOfTimeBoundsEntry.getDuplicateUntil)
+    val tooEarlyUntilMaybe = parseOptionalTimestamp(
+      outOfTimeBoundsEntry.hasTooEarlyUntil,
+      outOfTimeBoundsEntry.getTooEarlyUntil)
+    val tooLateFromMaybe = parseOptionalTimestamp(
+      outOfTimeBoundsEntry.hasTooLateFrom,
+      outOfTimeBoundsEntry.getTooLateFrom)
+    val deduplicated = duplicateUntilMaybe.exists(recordTime <= _)
+    val tooEarly = tooEarlyUntilMaybe.exists(recordTime <= _)
+    val tooLate = tooLateFromMaybe.exists(recordTime > _)
+    val invalidLedgerTime = tooEarly || tooLate
+
+    wrappedLogEntry.getPayloadCase match {
+      case _ if deduplicated =>
+        // We don't emit updates for deduplicated submissions.
+        None
```

Comment:
Right. AFAIK the ledger API server will get confused when it receives different updates for the same command ID. This will likely change in the future, but for now this seems to be the right behaviour.

Reply:
OK, thanks for the confirmation.